storage/blockdevice/COMPONENT_QSPIF/UNITTESTS/test_QSPIFBlockDevice.cpp

```diff
@@ -0,0 +1,262 @@
+/* Copyright (c) 2021 Arm Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+
+#include "QSPIF/QSPIFBlockDevice.h"
+
+// S25FS512S's CR3NV register needs a quirk
+static const int S25FS512S_CR3NV = 0x000004;
+
+namespace mbed {
+
+/**
+ * Fake implementation of mbed::QSPI class for QSPIFBlockDevice unit tests.
+ *
+ * Note: Data output by this is either dummy or based on Cypress S25FS512S.
+ * We can't use gMock for this because
+ * - The entire mbed::QSPI is non-virtual. Mocks are derived classes, and
+ * when a derived class instance is used as an instance of its base class,
+ * non-virtual members of the base class get usedx.
+ * - QSPIFBlockDevice's member _qspi (an instance of mbed::QSPI) is a value
+ * instead of a reference, in order to support initializing from pins directly.
+ * Also, mbed::QSPI is declared as NonCopyable whose copy-constructor is deleted,
+ * so we can't copy a preexisting mbed::QSPI instance into QSPIFBlockDevice.
+ * There's no way to dependency inject a mock instance.
```

Comment:
Could we change QSPIFBlockDevice's member to be a reference instead of a copy? This is an internal implementation detail and shouldn't affect use of the public API.

Reply:
The existing constructor is
```
QSPIFBlockDevice(PinName io0, PinName io1, PinName io2, PinName io3, PinName sclk, PinName csel,
                                   int clock_mode,
                                   int freq);
```
so to support this constructor, the member `QSPI _qspi` needs to be a "full" one (instead of a reference) in order to construct a QSPI object from pins. If we change this member to a reference `QSPI &_qspi`, then QSPIFBlockDevice won't reserve memory for it, and the workaround I can think of would be using heap to construct a QSPI object from pins in QSPIFBlockDevice's constructor.

Reply:
A pointer allow us to inject any QSPI object, no matter where constructed. Using a pointer seems to make injection easier after we also add a new constructor, but this is fine for now.

Reply:
If we add a new constructor that allows us to inject a `QSPI` object, we should probably take a `unique_ptr<QSPI>`. Passing a `unique_ptr` semantically means that we're taking ownership of the `QSPI` object and managing the memory in this class. 

If we just take a raw pointer in the constructor we're changing this class to be non-owning with respect to `QSPI`, which could be a relatively big change. We'd also have an "owning" and a "non-owning" constructor which could make the internal implementation more complex.

Reply:
Are QSPI objects themselves usually made on the heap anyway? Heap pressure should be about the same. I see your point about not wanting to add a dependency on the heap where we didn't have one before. Does QSPIFBlockDevice not use the heap in other places?

Reply:
The only direct use of the heap by QSPIFBlockDevice is the CSEL array. And the SFDP sector map parsing function temporarily stores SFDP data in the heap too, and it's freed when parsing finishes (the unique_ptr goes out of scope).